framework/aster-frame/src/cpu.rs

```diff
@@ -59,15 +146,27 @@ macro_rules! cpu_local {
 /// The `CpuLocal<T: Sync>` can be used directly.
 /// Otherwise, the `CpuLocal<T>` must be used through `CpuLocal::borrow_with`.
 ///
-/// TODO: re-implement `CpuLocal`
+/// On each CPU, the actual virtual address of the `CpuLocal` object is calculated as follows：
+/// Object VA = base address + offset
+/// FS register is used to store the base address of cpu local data, and the base addresses of
+/// different cores are different to ensure the privacy of data. The offset of the linked virtual
+/// address of this object from the start address of the `.cpu_local` segment is used as the
+/// offset of the actual virtual address from the base address.
+///
+/// Compared to directly accessing an object, `CpuLocal` requires some additional overhead to
+/// calculate the actual virtual address.
 pub struct CpuLocal<T>(UnsafeCell<T>);
```

Comment:
Days earlier I thought that `T` must implement `Pod` for it would be moved and assumed to be initialized. That's wrong since:

> Most values in Rust are entirely okay with being moved around at-will.

See [pinning](https://doc.rust-lang.org/std/pin/index.html). It justifies the seemingly unsafe usage of objects such as `Arc` or `Once` as part of `T`, as long as the alignment is satisfied after moving (which is satisfied since `__bsp_local_start` and `cpu_local_start` are base page aligned).

However, there's still some object that cannot be moved, which will implement `!Unpin`.

So I suggest `CpuLocal<T: core::marker::Unpin>` here.

Reply:
But we are copying instead of just moving around. Usages of things like `Arc` will still be broken if it's not being wrapped by an `Option`. But It seems that such usages cannot be avoided precisely at compile time.

Whether to let `cpu_local` be unsafe (remove the `unsafe {}` marker in the macro), or just to turn a blind eye, that's a problem.